### ISSUE TITLE
Master roadmap: concrete solutions for every current open issue + latent reasoning / multi-axis LM path

### ISSUE THREAD

**Issue by s-b-repo** [s-b-repo/unifur#4]
# Master solution roadmap for all current open issues

This issue consolidates the current open work in Unifur and turns it into an implementation + falsification roadmap. It is intentionally evidence-driven: mechanisms are hypotheses until benchmarked.

Current open issues covered:
- #1 `improvements`
- #2 `Research proposal: Hybrid MoVA + GLM-style sparse attention + Nemotron LatentMoE architecture`
- #3 `Research: Multi-axis sparse LM combining MoVA, GLM, Nemotron, and Unifur`

The repository already has substantial infrastructure for parallel depth denoising, consistency training, MoE, adaptive depth, loop graphs, quality gates, quantization/LoRA, distillation, profiling and numerical certificates. The remaining quality/reproduction work is largely GPU/cluster dependent. See `TODO.md` and `README.md`.

---

## 1. First principle: separate implementation from evidence

Every proposed feature should have four statuses:

- **Implemented** — code exists and tests pass.
- **Verified** — controlled experiments demonstrate the claimed benefit.
- **Plausible** — technically motivated but not yet demonstrated.
- **Rejected** — measured and shown not to help or to regress.

No theorem, speedup, quality claim or architecture should be called proven merely because the mechanism is elegant.

---

# A. Solutions for issue #1 — training, mathematics, quality and reproducibility

## A1. Finish the blocked GPU reproduction matrix

Run the baseline and Unifur variants on identical datasets, seeds and hardware.

Required comparisons:

1. original/sequential baseline
2. Unifur sequential
3. parallel K=2
4. parallel K=3
5. hybrid
6. adaptive
7. consistency variants
8. flow matching vs EDM
9. MoE expert counts
10. distillation
11. precision variants

Report both quality and total wall-clock/GPU-hours. A faster optimizer step is not enough.

## A2. Replace undocumented hyperparameter assumptions with sweeps

Automate reproducible sweeps for:

- learning rate
- optimizer parameters
- consistency weight
- overlap/gamma
- block count
- K parallelism
- noise/sigma schedule
- MoE expert count
- top-k
- adaptive-depth ponder penalty
- gate thresholds
- precision policy

Store every configuration and raw measurement.

## A3. Build a matched-budget benchmark harness

Every comparison should support matching on:

- parameter count
- active parameters/token or sample
- optimizer steps
- examples/tokens seen
- FLOPs
- wall-clock time
- GPU-hours

Produce Pareto curves instead of one headline score.

## A4. Fix the mathematical-claim problem by making assumptions explicit

For every theorem/bound in the documentation:

1. state assumptions exactly
2. check dimensions and norms
3. verify inequality directions
4. identify required smoothness/Lipschitz/convexity/boundedness assumptions
5. distinguish statements about an idealized objective from statements about neural-network optimization
6. downgrade unsupported theorems to propositions/heuristics
7. add machine-checkable residual or numerical certificates where appropriate

The strongest useful result is an actually applicable bound, not the strongest-sounding claim.

## A5. Empirically test error propagation through block composition

For each block record:

- local error
- boundary mismatch
- sensitivity/Jacobian proxy
- downstream amplification
- contribution to final error

Compare empirical propagation against the current theoretical model. Replace simplistic additive assumptions if experiments disagree.

## A6. Turn consistency loss into an experimental question

Run no-consistency, low, medium and high weights.

Measure:

- boundary agreement
- final quality
- gradient interference
- optimization stability
- compute overhead

Determine whether consistency improves the task or merely forces representations to agree.

## A7. Validate adaptive depth instead of trusting confidence

Record:

- layer/span usage
- skip rate
- early-exit accuracy
- false early exits
- calibration
- expected depth
- quality vs compute

Only keep adaptive depth if it moves the Pareto frontier.

## A8. Validate MoE routing health

Track:

- expert utilization
- routing entropy
- load imbalance
- expert collapse
- routing stability
- dispatch overhead

Add regularization/capacity controls only when supported by observed failure modes.

## A9. Make checkpoints fully reproducible

Persist:

- model
- optimizer
- scheduler
- scaler
- RNG state
- data cursor
- config
- git revision
- dependency/environment identity
- dataset identity

Add integrity checks and a resume equivalence test.

## A10. Finish the systems audit

Measure where time is actually spent:

- dataloader
- CPU preprocessing
- GPU kernels
- synchronization
- communication
- checkpoint I/O
- memory bandwidth
- expert dispatch

Optimize the dominant bottleneck, not whichever component sounds most sophisticated.

## A11. Treat io_uring/native precision/DDP as benchmark hypotheses

The README intentionally avoids several external-dependency/backend commitments. Keep the portable implementation, but add optional feature-gated experiments where hardware/backend permits.

Compare rather than assume:

- positional/coalesced reads vs native io_uring
- precision emulation vs native BF16/FP16
- single GPU vs DDP
- replicated experts vs sharded experts

---

# B. Solutions for issue #2 — Hybrid MoVA + sparse/recurrent attention + LatentMoE

## B1. Build a small language-model research branch first

Do not start with a 24–32B model. First implement a compact LM where all tensor shapes and routing paths can be exhaustively profiled.

Minimum prototype:

- dense transformer baseline
- hybrid attention module
- MoVA value routing
- latent FFN router
- optional persistent routing state

The first objective is architectural validity, not benchmark prestige.

## B2. Hybrid attention stack

Implement a configurable mixture of:

- efficient recurrent/linear attention for most layers
- periodic precise sparse retrieval attention
- standard dense attention as a control

The layer schedule must be configurable so 1:1, 2:1, 3:1, 4:1 and learned schedules can be compared.

## B3. Hierarchical MoVA

Instead of routing directly over a large flat expert set:

```text
hidden
  -> family router
  -> top-k families
  -> expert router
  -> top-k value experts
```

Start with small expert counts and scale only if routing remains stable.

## B4. Shared-base + low-rank value experts

Avoid maintaining a full independent value matrix for every expert.

Use:

```text
V_expert = V_shared + DeltaV_lowrank
```

Measure:

- parameter memory
- bandwidth
- latency
- quality
- expert specialization

This directly addresses the residency problem in large MoVA pools.

## B5. LatentMoE routing

Compress the representation used for routing before selecting a large FFN expert pool.

Keep FFN expertise and MoVA value expertise as separate spaces.

The token may select:

```text
FFN experts: A, B
MoVA experts: C, D
```

rather than a single monolithic expert identity.

## B6. Shared persistent routing state

Introduce a compact routing state:

```text
r_l = g(r_{l-1}, h_l)
```

Use it to condition:

- FFN routing
- MoVA routing
- sparse-attention activation
- adaptive depth
- memory access

Measure whether this reduces routing recomputation and improves locality without creating routing inertia.

## B7. Routing-locality/cache experiment

For long agentic trajectories, measure whether consecutive tokens/states repeatedly choose similar experts.

If yes, add cache-friendly expert execution and expert residency/paging policies.

If no, do not add the complexity.

---

# C. Solutions for issue #3 — Multi-axis sparse LM

## C1. Use explicit compute axes

Treat conditional compute as independent axes:

```text
knowledge      -> Engram/static memory
sequence       -> recurrent state
retrieval      -> sparse attention
value capacity -> MoVA
FFN capacity   -> LatentMoE
layer depth    -> adaptive depth
prediction     -> MTP
```

The system should allocate compute per token instead of simply maximizing global model size.

## C2. Multi-axis controller

Create a controller that predicts a compute policy from the hidden state:

```text
policy = P(h, r, budget)
```

Possible actions:

- attention mode
- memory read
- MoVA top-k
- FFN top-k
- extra depth
- verification
- MTP width

Initially implement this as a deterministic policy interface so each component can be ablated independently.

## C3. Adaptive depth for language

Port Unifur's existing adaptive-depth concepts to the LM path:

```text
easy token      -> skip / shallow
normal token    -> standard depth
hard token      -> additional latent blocks
critical token -> additional blocks + verification
```

Measure quality vs average compute rather than average depth alone.

## C4. Latent reasoning: separate compute steps from emitted tokens

This is the key new research direction from this issue set.

Instead of forcing every reasoning step to become a generated token:

```text
prompt
  -> latent state H0
  -> H1
  -> H2
  -> ...
  -> Hk
  -> token logits
```

Allow several internal computation/refinement steps before the next token is emitted.

The important experiment is whether reasoning quality scales with **latent compute** independently of visible/generated-token count.

## C5. Parallel latent branches

Adapt Unifur's parallel trajectory concept to latent language representations:

```text
          H0
      /    |    |    \
    B1     B2   B3     B4
      \    |    |    /
          aggregator
              |
             H1
```

Test aggregators:

- mean
- learned weighted sum
- attention
- confidence weighted
- gated mixture

## C6. Latent progress objective

Consistency alone can collapse blocks into identity mappings.

Add a progress objective that rewards a later latent state for improving the target prediction.

Conceptually:

```text
error(H_{i+1}) < error(H_i)
```

Use a margin-based or verifier-derived objective and test for genuine progress rather than representational similarity.

## C7. Internal verifier and refinement loop

Add a lightweight verifier over latent states:

```text
H
 -> provisional answer
 -> verifier
    -> accept
    -> refine
```

For difficult cases:

```text
H0 -> verify -> reject -> H1 -> verify -> ...
```

Limit refinement with a compute budget to prevent runaway inference.

## C8. Verification-aware early exit

Do not equate confidence with correctness.

Train and calibrate a verifier/halting signal against actual correctness.

Record false-positive early exits and false-negative continuation.

## C9. Adaptive MTP

Implement 2–4 shared-weight multi-token prediction heads as a separate acceleration axis.

Start with fixed-width MTP, then test confidence-conditioned widths:

```text
high confidence   -> wider proposal
medium confidence -> smaller proposal
low confidence    -> one-token proposal
```

Measure accepted speculative tokens and end-to-end latency.

## C10. Memory hierarchy

Use a clean division of responsibilities:

- static/frequent knowledge: conditional memory
- recent sequence state: recurrent/linear state
- precise retrieval: sparse attention
- long-term learned memory: optional learned memory path

Do not enable every memory mechanism in the same initial experiment. Add them one at a time.

---

# D. Cross-cutting solutions that apply to all three issues

## D1. Common router interface

Unifur should expose a shared routing abstraction capable of returning:

- selected experts
- gates
- entropy
- capacity/load statistics
- routing state
- compute cost

MoVA, FFN MoE and future memory/attention routers should use the same instrumentation even when their implementations differ.

## D2. Unified compute budget

Introduce a budget object representing:

- FLOPs
- memory
- latency target
- expert residency budget
- maximum latent reasoning steps

Every adaptive mechanism should consult this budget.

## D3. Unified quality gate

Extend the existing quality-gate philosophy from numerical correctness into LM experiments:

```text
metric -> threshold -> verdict
```

Possible verdicts:

- pass
- fail
- inconclusive

Include seed variance so noise is not mistaken for progress.

## D4. Standard benchmark matrix

For every model version capture:

- perplexity
- coding
- mathematical reasoning
- logical reasoning
- long-context retrieval
- tool/agent trajectories
- calibration
- robustness to distractors
- average active compute
- p50/p95 latency
- memory footprint
- throughput

Use identical evaluation settings across variants.

## D5. Required ablation order

The experiment order should be:

1. baseline
2. one mechanism
3. second mechanism
4. interaction of the two
5. full architecture

Never jump directly from baseline to a many-component architecture and attribute gains to every component.

## D6. Negative-result ledger

Record unsuccessful experiments with:

- configuration
- hypothesis
- result
- likely failure mode
- whether the component should be abandoned, modified or retained only as infrastructure

This prevents repeatedly rediscovering failed approaches.

---

# E. Suggested implementation sequence

### Stage 1 — evidence foundation

- complete GPU reproduction matrix
- add matched-budget benchmark harness
- finish mathematical audit
- standardize experiment recording

### Stage 2 — language substrate

- compact LM baseline
- transformer + Unifur routing abstractions
- parallel latent block API

### Stage 3 — sparse compute

- hybrid attention
- hierarchical MoVA
- latent FFN routing
- routing diagnostics

### Stage 4 — adaptive compute

- adaptive depth
- latent refinement
- verifier
- budget controller

### Stage 5 — memory/prediction

- conditional/static memory
- optional long-term memory
- MTP

### Stage 6 — scale-up

Only after the compact model shows a reproducible Pareto improvement:

- increase experts
- increase hidden size
- increase context
- increase model size
- measure distributed/expert-residency behavior

---

# F. Acceptance criteria

The project should not declare the combined architecture successful because it contains many modern components.

A successful result should demonstrate one or more of:

- better quality at fixed compute
- lower compute at fixed quality
- lower memory at fixed quality
- lower latency at fixed quality
- better long-context/agentic quality at fixed serving cost
- improved reasoning from latent computation without proportional visible reasoning tokens

and should include the ablations needed to identify the cause.

# Central research hypothesis

The most important new hypothesis is:

> **LLMs may benefit from separating internal computation from token emission: multiple learned latent refinement/verification steps can occur between emitted tokens, with Unifur providing the parallel, adaptive, consistency-regularized computation framework.**

This is a research hypothesis, not an established result. The first decisive experiment is a small language model comparing ordinary autoregressive depth against sequential and parallel latent refinement at matched compute and output length.

---

## Definition of done

This umbrella roadmap is complete when:

- each current open issue has either an implementation, a measured result, or a documented external blocker;
- every major optimization has an ablation;
- mathematical claims are classified correctly;
- raw benchmark data is reproducible;
- the best architecture is selected by measured quality/compute Pareto efficiency;
- and unsuccessful ideas remain documented rather than silently removed.

The governing rule is:

**Optimize verified quality per unit compute, not the number of architectural ideas implemented.**
